loader/loader.go

```diff
@@ -1,5 +1,209 @@
+	dbTx       *database.DBTx
+	data       T
+}
+
+type BlockRange struct {
+	startBlockHeight int64
+	endBlockHeight   int64
+}
+
+// NewLoaderManager creates a new LoaderManager and initiates goroutines for the loaders in a pipeline.
+func NewLoaderManager(ctx context.Context, client client.Client, db database.Database, opts LoaderOptions) (*LoaderManager, error) {
+	// initialize struct
+	ctx, cancel := context.WithCancel(ctx)
+	g, ctx := errgroup.WithContext(ctx)
+	inputCh := make(chan *LoaderMsg[BlockRange])
+	loader := &LoaderManager{
+		client:  client,
+		db:      db,
+		inputCh: inputCh,
+		ctx:     ctx,
+		cancel:  cancel,
+		g:       g,
+	}
+
+	// loaders
+	blockRangeLoader := NewLoader[BlockRange, []*chainhash.Hash](client, inputCh, blockRangeHandler)
+	g.Go(blockRangeLoader.Run)
+	blockHashLoader := NewLoader[[]*chainhash.Hash, []*types.Block](client, blockRangeLoader.Dst(), blockHashHandler)
+	g.Go(blockHashLoader.Run)
+	blockLoader := NewLoaderSink[[]*types.Block](blockHashLoader.Dst(), blockHandler)
+	g.Go(blockLoader.Run)
+
+	return loader, nil
+}
+
+// Close gracefully shuts down all loader processes.
+func (loader *LoaderManager) Close() error {
+	done := false
+	loader.stopOnce.Do(func() {
+		close(loader.inputCh)
+		loader.cancel()
+		done = true
+	})
+	if !done {
+		return fmt.Errorf("already closed")
+	}
+	return loader.g.Wait()
+}
+
+// SendInput starts the given parameters on the first stage of the loader pipeline.
+func (loader *LoaderManager) SendInput(blockRange BlockRange, dbTx database.DBTx) {
+	msg := &LoaderMsg[BlockRange]{
+		blockRange,
+		&dbTx,
+		blockRange,
+	}
+	loader.inputCh <- msg
+}
+
+// ILoader is a simple interface for loaders.
+type ILoader interface {
+	Run() error
+}
+
+// Loader is the go to loader for an inidividual stage in the pipeline
+// extracting data from an RPC client to a database. It stores state and
+// uses the function f to transform data coming from a src channel to send
+// to a dst channel.
+type Loader[S, D any] struct {
+	client client.Client
+	src    <-chan *LoaderMsg[S]
+	dst    chan *LoaderMsg[D]
+	f      LoaderFunc[S, D]
+}
+
+type LoaderFunc[S, D any] func(client.Client, *LoaderMsg[S]) (*LoaderMsg[D], error)
+
+func (loader *Loader[S, D]) Dst() <-chan *LoaderMsg[D] {
+	return loader.dst
+}
+
+func NewLoader[S, D any](client client.Client, src <-chan *LoaderMsg[S], f LoaderFunc[S, D]) *Loader[S, D] {
+	dst := make(chan *LoaderMsg[D])
+	loader := &Loader[S, D]{
+		client,
+		src,
+		dst,
+		f,
+	}
+	return loader
+}
+
+// Run listens for messages sent to the loader's src channel,
+// transforms the data and sends it to the next loader.
+// Importantly, when it's src channel is closed, it closes its
+// dst channel, which causes a domino effect of closing loader channels.
+func (loader *Loader[S, D]) Run() error {
+	for msg := range loader.src {
+		output, err := loader.f(loader.client, msg)
+		if err != nil {
+			return err
+		}
+		loader.dst <- output
+	}
+	close(loader.dst)
```

Comment:
```suggestion
	defer close(loader.dst)
	for msg := range loader.src {
		output, err := loader.f(loader.client, msg)
		if err != nil {
			return err
		}
		loader.dst <- output
	}
```

Btw I think it's best to use defer here to always make sure we close the dst channel of this loader whenever execution stops for any reason. Before, if an error occurred, the channel would be left open.